### PR TITLE
added an alert system for expiring covert actions and assualt mission…

### LINKED ISSUE #382: Create optional alert when an activity is about to expire
The alert would be either a popup window or a side notification that pauses the passage of time. It would be controlled from MCM, with an option to enable or disable it.

### PR DESCRIPTION
**Pull request by statusNone**
closes #382

### CODE REVIEW

File: CovertInfiltration/Config/XComModConfigMenu_Defaults.ini [hunk 1]
```diff
@@ -8,4 +8,7 @@ ENABLE_TUTORIAL_DEFAULT=true
 
 REMOVE_NICKNAMED_UPGRADES_DEFAULT=true
 
+WARN_BEFORE_EXPIRATION_DEFAULT=true
```
Comment (Xymanek) [inline CovertInfiltration/Config/XComModConfigMenu_Defaults.ini]: Should be off by default - it's an opt-in feature for those who want it

File: CovertInfiltration/Config/XComModConfigMenu_Defaults.ini [hunk 2]
```diff
@@ -8,4 +8,7 @@ ENABLE_TUTORIAL_DEFAULT=true
...
+HOURS_BEFORE_WARNING_DEFAULT=2
```
Comment (Xymanek) [inline CovertInfiltration/Config/XComModConfigMenu_Defaults.ini]: I would make it 3 hours by default

File: CovertInfiltration/Localization/CovertInfiltration.int [hunk 3]
```diff
@@ -83,6 +83,10 @@ EnableTutorialDesc="Enable tutorial"
 EnableTutorialTooltip="Show in-game explanations and tips regarding changes and new features/mechanics"
 RemoveNicknamedUpgradesDesc="Remove upgrades from nicknamed items"
 RemoveNicknamedUpgradesTooltip="Disabling this will preserve any nicknamed weapon's upgrades when selecting 'MAKE UPGRADES AVAILABLE' at the squad select screen"
+WarnBeforeExpirationDesc="Warn Before Expiration"
+WarnBeforeExpirationTooltip="Warn before an activity is about to expire"
+HoursBeforeWarningDesc="Hours Before Warning"
+HoursBeforeWarningTooltip="How many hours left before warning"
```
Comment (Xymanek) [inline CovertInfiltration/Localization/CovertInfiltration.int]: Not thrilled about wording, but can't really say what's wrong with it....

cc @cannonfodder1 wdyt?

File: CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_Activity_Assault.uc [hunk 4]
```diff
@@ -36,4 +46,22 @@ protected function UpdateActivity ()
...
 	}
+	else if (WarnBeforeExpiration && !bAlreadyWarnedOfExpiration)
+	{
+		HoursBeforeWarning = `MCM_CH_GetValueStatic(class'ModConfigMenu_Defaults'.default.HOURS_BEFORE_WARNING_DEFAULT, class'UIListener_ModConfigMenu'.default.HOURS_BEFORE_WARNING);
+		AdjustedTime = GetCurrentTime();
+		class'X2StrategyGameRulesetDataStructures'.static.AddHours(AdjustedTime, HoursBeforeWarning);
+
+		if (class'X2StrategyGameRulesetDataStructures'.static.LessThan(ExpiryTimerEnd, AdjustedTime))
+		{
+			NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: Activity" @ m_TemplateName @ "(assault mission) has expired");
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_Activity_Assault.uc]: Adjust the name of the state frame please

File: CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_Activity_Assault.uc [hunk 5]
```diff
@@ -36,4 +46,22 @@ protected function UpdateActivity ()
 
 		`SubmitGameState(NewGameState);
 	}
+	else if (WarnBeforeExpiration && !bAlreadyWarnedOfExpiration)
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_Activity_Assault.uc]: You need to check `bExpiring` here as well

File: CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_CovertActionExpirationManager.uc [hunk 6]
```diff
@@ -124,4 +151,18 @@ function AddActionExpirationInfo(StateObjectReference ActionRef, TDateTime Expir
...
+
+function WarnOfExpiration(StateObjectReference WarningRef, XComGameState_CovertAction CovertAction)
+{
+	local int idx;
+	
+	for (idx = 0; idx < ActionExpirationInfoList.Length; idx++)
+	{
+		if (ActionExpirationInfoList[idx].ActionRef == WarningRef)
+		{
+			ActionExpirationInfoList[idx].bAlreadyWarnedOfExpiration = true;
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_CovertActionExpirationManager.uc]: I like how you comment "can't have an open gamestate" but then edit a state variable here

File: CovertInfiltration/Src/CovertInfiltration/Classes/UIUtilities_Infiltration.uc [hunk 7]
```diff
@@ -304,6 +306,11 @@ static function AssaultMissionAvaliable (XComGameState_MissionSite MissionState)
 	GeoscapeOpportunityAvaliable(MissionState, default.strAssaultAvaliableHeader, MissionState.GetUIButtonIcon(), MissionState.GetMissionObjectiveText());
 }
 
+static function AssaultMissionExpiring (XComGameState_MissionSite MissionState)
+{
+	GeoscapeOpportunityAvaliable(MissionState, default.strAssaultExpiringHeader, MissionState.GetUIButtonIcon(), MissionState.GetMissionObjectiveText(), eUIState_Warning);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/UIUtilities_Infiltration.uc]: Not a fan of calling `GeoscapeOpportunityAvaliable` here as the name/purpose obviously clashes. If it's suitable, then perhaps rename it to `GeoscapeOpportunityNotification`?

File: CovertInfiltration/Localization/CovertInfiltration.int (hunk 3 above)
Comment (cannonfodder1) [reply CovertInfiltration/Localization/CovertInfiltration.int]: Titles are fine the tooltips need some tiny changes.
"Show a warning before an activity is about to expire"
"The warning will appear this many hours before expiry"

File: CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_CovertActionExpirationManager.uc (hunk 6 above)
Comment (statusNone) [reply CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_CovertActionExpirationManager.uc]: >.>

**Pull request merged**